Classifier: Set step.next for transcription tasks

eatyourgreens opened the pull request:
Transcription workflows have steps defined, so they bypass `convertWorkflowToUseSteps`. However, the Next Task selector in the project builder still sets `task.next` to a task key for each task and single choice answer. This PR runs `convertWorkflowToUseSteps` on all incoming workflow snapshots. It checks `step.next` for each step, and tries to set it from `task.next` if it doesn't exist.

https://localhost:8080/?env=production&project=9006&workflow=18244&demo=true

Package:
lib-classifier

Closes #2102.

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

eatyourgreens commented:
Looking at this again, maybe it would be better to put these fixes into `convertWorkflowToUseSteps` and run that across all workflows, even ones that do have `steps` defined.

srallen reviewed (changes requested):
Blocking this because I think the fix should go into the lab. When the new lab page is built, would this be removed? I'd prefer not editing the classifier for a scenario of configuration that shouldn't exist to begin with.

eatyourgreens commented:
I think the plan is to remove `convertWorkflowToUseSteps` eventually. It was originally part of the `WorkflowSteps` model and my understanding is it's only needed for workflows that are expected to work in PFE?

srallen commented:
Ok, for posterity, after I got more fully caught up, the issue for #2102 isn't how the lab works, but with the slightly differing behavior between `AnnotatedSteps` and `WorkflowStepStore`. I've opened #2107 as an alternative fix to this because it addresses the differing behaviors and refactors `step.getNextStep` to derive the next step using the way `WorkflowStepStore` built it to do so.

eatyourgreens commented:
I'm afraid #2107 introduces a regression that was fixed in #2097 for branching workflows. That bug isn't present on this branch. The missing piece in #2107 is a check on each step to see if it's the last step the workflow. Branching workflows can have more than one last step.

I started down the same route here, initially. The first version of this PR modified the `WorkflowSteps` model to check each step for a next step. However, I realised that I was duplicating code from the `Workflow` model, and the fix was simpler if I added ~5 new lines to `convertWorkflowToUseSteps`, so I rewrote this PR and went down that route instead.

srallen commented on packages/lib-classifier/src/store/helpers/convertWorkflowToUseSteps/convertWorkflowToUseSteps.js:
```diff
@@ -97,8 +97,21 @@ export default function convertWorkflowToUseSteps ({ first_task, tasks }) {
+}
+
+export default function convertWorkflowToUseSteps ({ first_task, steps = [], tasks }) {
+  steps = steps.length > 0 ? steps :createStepsFromTasks(first_task, tasks)
```
```suggestion
  steps = steps.length > 0 ? steps : createStepsFromTasks(first_task, tasks)
```

srallen reviewed (approved):
There's a typo, but it doesn't seem to be breaking anything.
